<a href="https://colab.research.google.com/github/iekim9/Iris-Anova-KMeans/blob/master/tf_idf_with_logistic_regression.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import numpy as np
import nltk
import re
from sqlalchemy import create_engine

In [2]:
engine = create_engine('postgresql+psycopg2://postgres:Stats170AB@stats170.c56hn1tg3mkg.us-east-1.rds.amazonaws.com:5432')

/usr/local/lib/python3.7/dist-packages/psycopg2/__init__.py:144: UserWarning: The psycopg2 wheel package will be renamed from release 2.8; in order to keep installing from binary please use "pip install psycopg2-binary" instead. For details see: <http://initd.org/psycopg/docs/install.html#binary-install-from-pypi>.
  """)


In [3]:
query = "SELECT * from fullDocuments f join documentLabels d on f.docid = d.docid;"
fulldocs = pd.read_sql_query(query, con=engine)
fulldocs

docid                                           fulltext  ...  caseid  label
0      1.0  Document Text: Example 1\nReferring Doctor: Un...  ...      -1    1.0
1      2.0  Document Text: Example 2\nProgress Notes\nDate...  ...      -1    1.0
2      3.0  Document Text: Example 3\nChief Complaint:\n1....  ...      -1    4.0
3      4.0  Document Text: 14\nse\nLAIDA\n*\nPatient Infor...  ...   11594    1.0
4      5.0  Document Text: The first set of documents behi...  ...   11594    1.0
..     ...                                                ...  ...     ...    ...
110  111.0  Document Text: 9:10\nFAX\n2002/041\n'\nPatient...  ...   87832    1.0
111  112.0  Document Text: # 23/24\nfined VI\nvyy\nScan on...  ...   88178    1.0
112  113.0  Document Text: 15:03 From\nPage: 1/24\nNUFACTO...  ...   95139    1.0
113  114.0  Document Text: 03:58:42 p.m.\n1/2\nNU FACTOR\n...  ...   95739    2.0
114  115.0  Document Text: No. 0902 P. 1\nIV/IG Protocol\n...  ...   97245    1.0

[115 rows x 5 columns]

In [4]:
corpus_dict = {}
corpus = []
tags = []
for i, row in fulldocs.iterrows():
  corpus.append(row['fulltext'])
  tags.append(row['label'])
  corpus_dict[i] = (row['fulltext'], row['label'])

In [5]:
from sklearn.model_selection import train_test_split
features_train, features_test, labels_train, labels_test = train_test_split(corpus, tags, test_size=0.2, random_state=10)
from sklearn.feature_extraction.text import TfidfVectorizer
vectorizer = TfidfVectorizer(stop_words='english')
features_train = vectorizer.fit_transform(features_train)
features_test = vectorizer.transform(features_test)

In [6]:
from sklearn.feature_selection import SelectPercentile, f_classif
selector = SelectPercentile(f_classif, percentile=10)
selector.fit(features_train, labels_train)
features_train = selector.transform(features_train).toarray()
features_test = selector.transform(features_test).toarray()

In [7]:
from sklearn.linear_model import LogisticRegression
# from sklearn.model_selection import cross_val_score, StratifiedKFold

In [8]:
model = LogisticRegression(C=100).fit(features_train, labels_train)
score = model.score(features_test, labels_test)
print(score)

0.4782608695652174


Adjusting C, which determines the amount of regularization.
Lower values increase regularization

In [9]:
from sklearn.model_selection import GridSearchCV

In [10]:
param_grid_ = {'C': [1e-5, 1e-3, 1e-1, 1e0, 1e1, 1e2]}
tfidf_search = GridSearchCV(LogisticRegression(), cv=5, param_grid=param_grid_)

In [12]:
Cs = [1e-7, 1e-5, 1e-3, 1e-1, 1e0, 1e1, 1e2, 1e3, 1e4, 1e5]

for c in Cs:
  curr_model = LogisticRegression(C=c).fit(features_train, labels_train)
  curr_score = curr_model.score(features_test, labels_test)
  print(f'{c}: {curr_score}')

1e-07: 0.43478260869565216
1e-05: 0.43478260869565216
0.001: 0.43478260869565216
0.1: 0.43478260869565216
1.0: 0.43478260869565216
10.0: 0.4782608695652174
100.0: 0.4782608695652174
1000.0: 0.4782608695652174


/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)
/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression


10000.0: 0.4782608695652174
100000.0: 0.4782608695652174


highest accuracy:
max_df = 0.5 or without this, C = 100, 0.4782608695652174

considering different parameters:
- penalty
- C
- class_weight
- solver: 'newton-cg', 'sag', 'saga', 'lbfgs'
- max_iter

with bigrams

In [13]:
from sklearn.model_selection import train_test_split
features_train, features_test, labels_train, labels_test = train_test_split(corpus, tags, test_size=0.2, random_state=10)

In [14]:
from sklearn.feature_extraction.text import TfidfVectorizer
vectorizer = TfidfVectorizer(stop_words='english', ngram_range = (1,2))
features_train = vectorizer.fit_transform(features_train)
features_test = vectorizer.transform(features_test)

In [15]:
for c in Cs:
  curr_model = LogisticRegression(C=c).fit(features_train, labels_train)
  curr_score = curr_model.score(features_test, labels_test)
  print(f'{c}: {curr_score}')

1e-07: 0.43478260869565216
1e-05: 0.43478260869565216
0.001: 0.43478260869565216
0.1: 0.43478260869565216
1.0: 0.43478260869565216
10.0: 0.43478260869565216
100.0: 0.4782608695652174
1000.0: 0.4782608695652174
10000.0: 0.4782608695652174
100000.0: 0.4782608695652174


with penalty l1

In [16]:
l1_model = LogisticRegression(penalty = 'l1', solver = 'saga').fit(features_train, labels_train)
l1_score = l1_model.score(features_test, labels_test)
print(l1_score)

0.43478260869565216


In [18]:
Cs = [1e-7, 1e-5, 1e-3, 1e-1, 1e0, 1e1, 1e2, 1e3, 1e4, 1e5]

for c in Cs:
  curr_model = LogisticRegression(penalty = 'l1', solver = 'saga', C=c).fit(features_train, labels_train)
  curr_score = curr_model.score(features_test, labels_test)
  print(f'{c}: {curr_score}')

1e-07: 0.43478260869565216
1e-05: 0.43478260869565216
0.001: 0.43478260869565216
0.1: 0.43478260869565216
1.0: 0.43478260869565216


/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_sag.py:330: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)


10.0: 0.4782608695652174


/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_sag.py:330: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)


100.0: 0.4782608695652174


/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_sag.py:330: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)


1000.0: 0.4782608695652174


/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_sag.py:330: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)


10000.0: 0.4782608695652174
100000.0: 0.43478260869565216


/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_sag.py:330: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)


stratified kfold split

In [19]:
from sklearn.model_selection import StratifiedKFold
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics import accuracy_score, f1_score

vectorizer = TfidfVectorizer(stop_words='english')

In [20]:
skf = StratifiedKFold(n_splits=5, shuffle=True, random_state=1)

In [21]:
corpus_np = np.array(corpus)
tags_np = np.array(tags)

In [23]:
vectorizer = TfidfVectorizer(stop_words='english')
accuracy_scores = []
Cs = [1e-7, 1e-5, 1e-3, 1e-1, 1e0, 1e1, 1e2, 1e3, 1e4, 1e5]

for train_index, test_index in skf.split(corpus, tags):
  features_train, features_test = corpus_np[train_index], corpus_np[test_index]
  labels_train, labels_test = tags_np[train_index], tags_np[test_index]
  features_train = vectorizer.fit_transform(features_train)
  features_test = vectorizer.transform(features_test)
  current = []


  for c in Cs:
    curr_model = LogisticRegression(penalty = 'l1', solver = 'saga', C=c).fit(features_train, labels_train)
    curr_score = curr_model.score(features_test, labels_test)
    # print(f'{c}: {curr_score}')
    current.append(curr_score)

  accuracy_scores.append(current)

/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_sag.py:330: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)
/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_sag.py:330: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)
/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_sag.py:330: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)
/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_sag.py:330: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)
/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_sag.py:330: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the 

In [25]:
from collections import defaultdict

In [29]:
skf = StratifiedKFold(n_splits=6, shuffle=True, random_state=1)
vectorizer = TfidfVectorizer(stop_words='english')
accuracy_scores = []
Cs = [1e-7, 1e-5, 1e-3, 1e-1, 1e0, 1e1, 1e2, 1e3, 1e4, 1e5]

train_dict = defaultdict(list)
test_dict = defaultdict(list)
f1_train_dict = defaultdict(list)
f1_test_dict = defaultdict(list)

for train_index, test_index in skf.split(corpus, tags):
  features_train, features_test = corpus_np[train_index], corpus_np[test_index]
  labels_train, labels_test = tags_np[train_index], tags_np[test_index]
  features_train = vectorizer.fit_transform(features_train)
  features_test = vectorizer.transform(features_test)
  current = []

  for c in Cs:
    curr_model = LogisticRegression(penalty = 'l1', solver = 'saga', C=c).fit(features_train, labels_train)
    # curr_score = curr_model.score(features_test, labels_test)
    # current.append(curr_score)

    labels_pred = curr_model.predict(features_test)
    labels_train_pred = curr_model.predict(features_train)

    train_dict[c].append(accuracy_score(labels_train, labels_train_pred))
    test_dict[c].append(accuracy_score(labels_test, labels_pred))
    f1_train_dict[c].append(f1_score(labels_train, labels_train_pred, average = 'weighted'))
    f1_test_dict[c].append(f1_score(labels_test, labels_pred, average = 'weighted'))

#   accuracy_scores.append(current)
# accuracy_scores

/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_sag.py:330: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)
/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_sag.py:330: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)
/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_sag.py:330: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)
/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_sag.py:330: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)
/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_sag.py:330: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the 

In [30]:
for key in test_dict:
  print("Key: ", key, " Values: ", str(f1_test_dict[key]))

Key:  1e-07  Values:  [0.4499999999999999, 0.4889643463497453, 0.043062200956937795, 0.4889643463497453, 0.4245614035087719, 0.4245614035087719]
Key:  1e-05  Values:  [0.4499999999999999, 0.4889643463497453, 0.4889643463497453, 0.4889643463497453, 0.4245614035087719, 0.4245614035087719]
Key:  0.001  Values:  [0.4499999999999999, 0.4889643463497453, 0.4889643463497453, 0.4889643463497453, 0.4245614035087719, 0.4245614035087719]
Key:  0.1  Values:  [0.4499999999999999, 0.4889643463497453, 0.4889643463497453, 0.4889643463497453, 0.4245614035087719, 0.4245614035087719]
Key:  1.0  Values:  [0.4499999999999999, 0.4889643463497453, 0.4889643463497453, 0.4889643463497453, 0.4245614035087719, 0.4245614035087719]
Key:  10.0  Values:  [0.47857142857142854, 0.43557168784029043, 0.43557168784029043, 0.46315789473684216, 0.5491506544138123, 0.6120857699805068]
Key:  100.0  Values:  [0.44444444444444436, 0.4210526315789473, 0.4889643463497453, 0.4889643463497453, 0.5491506544138123, 0.499025341130604

In [31]:
mean_accuracy_scores = defaultdict(list)

for key in train_dict:
  mean_accuracy_scores['test'].append(np.mean([np.mean(i) for i in test_dict[key]]))
  mean_accuracy_scores['train'].append(np.mean([np.mean(i) for i in train_dict[key]]))
  mean_accuracy_scores['f1 test'].append(np.mean([np.mean(i) for i in f1_train_dict[key]]))
  mean_accuracy_scores['f1 train'].append(np.mean([np.mean(i) for i in f1_train_dict[key]]))

In [32]:
mean_accuracy_scores

defaultdict(list,
            {'f1 test': [0.39466984526656956,
              0.4606491875617013,
              0.4606491875617013,
              0.4606491875617013,
              0.4606491875617013,
              0.9283908694583202,
              1.0,
              1.0,
              1.0,
              1.0],
             'f1 train': [0.39466984526656956,
              0.4606491875617013,
              0.4606491875617013,
              0.4606491875617013,
              0.4606491875617013,
              0.9283908694583202,
              1.0,
              1.0,
              1.0,
              1.0],
             'test': [0.5298245614035088,
              0.6087719298245614,
              0.6087719298245614,
              0.6087719298245614,
              0.6087719298245614,
              0.5828947368421052,
              0.5745614035087719,
              0.6096491228070174,
              0.6008771929824561,
              0.6096491228070174],
             'train': [0.5392543859649123,
   

In [ ]:
mean_test_acc = []
mean_train_acc = []
mean_test_f1 = []
mean_train_f1 = []
for key in train_dict:
  mean_test_acc.append(np.mean([np.mean(i) for i in test_dict[key]]))
  mean_train_acc.append(np.mean([np.mean(i) for i in train_dict[key]]))
  mean_test_f1.append(np.mean([np.mean(i) for i in test_f1_dict[key]]))
  mean_train_f1.append(np.mean([np.mean(i) for i in train_f1_dict[key]]))